core/sql/optimizer/RelExpr.cpp

```diff
@@ -9055,19 +9056,33 @@ void Scan::setTableAttributes(CANodeId nodeId)
   setBaseCardinality(MIN_ONE (tableDesc->getNATable()->getEstRowCount())) ;
 }
 
-NABoolean Scan::updateableIndex(IndexDesc *idx)
+NABoolean Scan::updateableIndex(IndexDesc *idx, ValueIdSet& preds, 
+                                NABoolean & needsHalloweenProtection)
 {
   //
-  // Returns TRUE if the index (idx) can be used for a scan during 
-  // an UPDATE. Halloween problem is protected with a sort using
-  // Scan::requiresHalloweenForUpdateUsingIndexScan(). 
-  // Returns FALSE only for certain embedded updates now.
+  // Returns TRUE if the index (idx) can be used for a scan during an UPDATE.
+  // Returns TRUE with needsHalloweenProtection also set to TRUE, if the index
+  // needs a blocking sort for Halloween protection
+  // Otherwise, returns FALSE to prevent use of this index. 
+  // Using the index in this case requires Halloween protection, but is likely 
+  // inefficient since there are no useful preds on the index key columns, so
+  // we don't add this index to list of candidates.
   //
 
-  if (!getGroupAttr()->isEmbeddedUpdate())
-    return TRUE ;
-
+  // The conditions of when this index returns TRUE can also be expressed as
+  // returns true for an index if it is 
+  // a) a unique/clustering index or 
+  // b) has a unique predicate on its key or 
+  // c) has an equals or range predicate on all the index columns that 
```

Comment:
In the process of mentally going through these conditions (see the previous messages above) and by talking to Suresh, I finally came to realize that what I was thinking matches what Suresh says in the comment here. Sorry that it took me a long time to arrive at this conclusion.